syntax = "proto3";

package core;

// ============ СЕРВИС ============

service CoreService {
    // Преподаватели
    rpc CreateTeacher(CreateTeacherRequest) returns (TeacherResponse);
    rpc GetTeacher(GetTeacherRequest) returns (TeacherResponse);
    rpc UpdateTeacher(UpdateTeacherRequest) returns (TeacherResponse);
    rpc DeleteTeacher(DeleteTeacherRequest) returns (DeleteResponse);
    rpc ListTeachers(ListTeachersRequest) returns (TeachersListResponse);
    rpc SearchTeachers(SearchRequest) returns (TeachersListResponse);
    
    // Предпочтения преподавателей ⭐ KEY FEATURE
    rpc GetTeacherPreferences(GetPreferencesRequest) returns (PreferencesResponse);
    rpc SetTeacherPreferences(SetPreferencesRequest) returns (SetPreferencesResponse);
    rpc UpdatePreference(UpdatePreferenceRequest) returns (PreferenceResponse);
    rpc ClearPreferences(ClearPreferencesRequest) returns (ClearResponse);
    rpc GetAllPreferences(GetAllPreferencesRequest) returns (AllPreferencesResponse);
    
    // Группы
    rpc CreateGroup(CreateGroupRequest) returns (GroupResponse);
    rpc GetGroup(GetGroupRequest) returns (GroupResponse);
    rpc UpdateGroup(UpdateGroupRequest) returns (GroupResponse);
    rpc DeleteGroup(DeleteGroupRequest) returns (DeleteResponse);
    rpc ListGroups(ListGroupsRequest) returns (GroupsListResponse);
    
    // Студенты
    rpc CreateStudent(CreateStudentRequest) returns (StudentResponse);
    rpc GetStudent(GetStudentRequest) returns (StudentResponse);
    rpc UpdateStudent(UpdateStudentRequest) returns (StudentResponse);
    rpc DeleteStudent(DeleteStudentRequest) returns (DeleteResponse);
    rpc ListStudents(ListStudentsRequest) returns (StudentsListResponse);
    rpc GetGroupStudents(GroupStudentsRequest) returns (StudentsListResponse);
    
    // Дисциплины
    rpc CreateDiscipline(CreateDisciplineRequest) returns (DisciplineResponse);
    rpc GetDiscipline(GetDisciplineRequest) returns (DisciplineResponse);
    rpc ListDisciplines(ListDisciplinesRequest) returns (DisciplinesListResponse);
    
    // Учебная нагрузка
    rpc CreateCourseLoad(CreateCourseLoadRequest) returns (CourseLoadResponse);
    rpc GetCourseLoad(GetCourseLoadRequest) returns (CourseLoadResponse);
    rpc ListCourseLoads(ListCourseLoadsRequest) returns (CourseLoadsListResponse);
    rpc DeleteCourseLoad(DeleteCourseLoadRequest) returns (DeleteResponse);
    
    // Импорт нагрузки из Excel
    rpc ImportCourseLoads(ImportRequest) returns (ImportResponse);
    rpc GetImportStatus(ImportStatusRequest) returns (ImportStatusResponse);
    rpc GetImportBatches(ImportBatchesRequest) returns (ImportBatchesResponse);
    
    // Связь с пользователями (ms-auth)
    rpc LinkTeacherToUser(LinkRequest) returns (LinkResponse);
    rpc LinkStudentToUser(LinkRequest) returns (LinkResponse);
    rpc GetTeacherByUserId(UserIdRequest) returns (TeacherResponse);
    rpc GetStudentByUserId(UserIdRequest) returns (StudentResponse);
    
    // Health Check
    rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
}

// ============ МОДЕЛИ ============

message Teacher {
    int32 id = 1;
    string full_name = 2;
    string first_name = 3;
    string last_name = 4;
    string middle_name = 5;
    
    string email = 6;
    string phone = 7;
    
    string employment_type = 8;                // 'external', 'graduate', 'internal', 'staff'
    int32 priority = 9;                        // 1-4 (автоматически)
    
    string position = 10;
    string academic_degree = 11;
    string department = 12;
    
    int32 user_id = 13;
    
    bool is_active = 14;
    string hire_date = 15;
    string termination_date = 16;
    
    string created_at = 17;
    string updated_at = 18;
    
    // Дополнительно (при запросе)
    PreferencesInfo preferences_info = 19;
}

message PreferencesInfo {
    int32 total_preferences = 1;
    int32 preferred_slots = 2;
    float preferences_coverage = 3;            // % от всех слотов
}

message TeacherPreference {
    int32 id = 1;
    int32 teacher_id = 2;
    
    int32 day_of_week = 3;
    int32 time_slot = 4;
    
    bool is_preferred = 5;
    string preference_strength = 6;
    string reason = 7;
    
    string created_at = 8;
    string updated_at = 9;
}

message Group {
    int32 id = 1;
    string name = 2;
    string short_name = 3;
    
    int32 year = 4;
    int32 semester = 5;
    int32 size = 6;
    
    string program_code = 7;
    string program_name = 8;
    string specialization = 9;
    string level = 10;
    
    int32 curator_teacher_id = 11;
    string curator_name = 12;
    
    bool is_active = 13;
    string enrollment_date = 14;
    string graduation_date = 15;
    
    string created_at = 16;
    string updated_at = 17;
}

message Student {
    int32 id = 1;
    string full_name = 2;
    string first_name = 3;
    string last_name = 4;
    string middle_name = 5;
    
    string student_number = 6;
    
    int32 group_id = 7;
    string group_name = 8;
    
    string email = 9;
    string phone = 10;
    
    int32 user_id = 11;
    
    string status = 12;
    string enrollment_date = 13;
    
    string created_at = 14;
    string updated_at = 15;
}

message Discipline {
    int32 id = 1;
    string name = 2;
    string short_name = 3;
    string code = 4;
    
    string department = 5;
    int32 credit_units = 6;
    string discipline_type = 7;
    
    bool is_active = 8;
    string created_at = 9;
}

message CourseLoad {
    int32 id = 1;
    
    int32 discipline_id = 2;
    string discipline_name = 3;
    string discipline_code = 4;
    
    int32 teacher_id = 5;
    string teacher_name = 6;
    int32 teacher_priority = 7;
    
    int32 group_id = 8;
    string group_name = 9;
    int32 group_size = 10;
    
    string lesson_type = 11;
    int32 hours_per_semester = 12;
    int32 weeks_count = 13;
    int32 lessons_per_week = 14;
    
    int32 semester = 15;
    string academic_year = 16;
    
    string required_classroom_type = 17;
    int32 min_classroom_capacity = 18;
    
    bool is_active = 19;
    string source = 20;
    string import_batch_id = 21;
    
    string created_at = 22;
}

message ImportBatch {
    int32 id = 1;
    string batch_id = 2;
    
    string filename = 3;
    int32 file_size = 4;
    int32 semester = 5;
    string academic_year = 6;
    
    int32 total_rows = 7;
    int32 successful_rows = 8;
    int32 failed_rows = 9;
    repeated string errors = 10;
    
    string status = 11;
    
    string started_at = 12;
    string completed_at = 13;
    string imported_by_name = 14;
}

// ============ ЗАПРОСЫ ============

// --- TEACHERS ---

message CreateTeacherRequest {
    string full_name = 1;
    string first_name = 2;
    string last_name = 3;
    string middle_name = 4;
    
    string email = 5;
    string phone = 6;
    
    string employment_type = 7;                // required
    
    string position = 8;
    string academic_degree = 9;
    string department = 10;
    
    string hire_date = 11;
    int32 created_by = 12;
}

message GetTeacherRequest {
    oneof identifier {
        int32 id = 1;
        string email = 2;
        int32 user_id = 3;
    }
    
    bool include_preferences = 4;              // Включить статистику предпочтений
}

message UpdateTeacherRequest {
    int32 id = 1;
    
    // Все поля optional
    string full_name = 2;
    string email = 3;
    string phone = 4;
    string employment_type = 5;
    string position = 6;
    string department = 7;
    bool is_active = 8;
    
    int32 updated_by = 9;
}

message DeleteTeacherRequest {
    int32 id = 1;
    bool hard_delete = 2;
}

message ListTeachersRequest {
    int32 page = 1;
    int32 page_size = 2;
    
    // Фильтры
    repeated string employment_types = 3;
    repeated int32 priorities = 4;
    string department = 5;
    bool only_active = 6;
    
    string sort_by = 7;
    string sort_order = 8;
}

message SearchRequest {
    string query = 1;
    int32 limit = 2;
}

message TeacherResponse {
    Teacher teacher = 1;
    string message = 2;
}

message TeachersListResponse {
    repeated Teacher teachers = 1;
    int32 total_count = 2;
    int32 page = 3;
    int32 page_size = 4;
}

// --- PREFERENCES ⭐ ---

message GetPreferencesRequest {
    int32 teacher_id = 1;
}

message PreferencesResponse {
    int32 teacher_id = 1;
    string teacher_name = 2;
    int32 teacher_priority = 3;
    
    repeated TeacherPreference preferences = 4;
    
    // Статистика
    int32 total_preferences = 5;
    int32 preferred_count = 6;
    int32 not_preferred_count = 7;
    float coverage_percentage = 8;
}

message SetPreferencesRequest {
    int32 teacher_id = 1;
    repeated PreferenceItem preferences = 2;
    bool replace_existing = 3;                 // true = удалить старые
}

message PreferenceItem {
    int32 day_of_week = 1;
    int32 time_slot = 2;
    bool is_preferred = 3;
    string preference_strength = 4;
    string reason = 5;
}

message SetPreferencesResponse {
    bool success = 1;
    int32 created_count = 2;
    int32 updated_count = 3;
    int32 deleted_count = 4;
    string message = 5;
}

message UpdatePreferenceRequest {
    int32 teacher_id = 1;
    int32 day_of_week = 2;
    int32 time_slot = 3;
    
    bool is_preferred = 4;
    string preference_strength = 5;
    string reason = 6;
}

message PreferenceResponse {
    TeacherPreference preference = 1;
    string message = 2;
}

message ClearPreferencesRequest {
    int32 teacher_id = 1;
}

message ClearResponse {
    bool success = 1;
    int32 deleted_count = 2;
}

message GetAllPreferencesRequest {
    int32 semester = 1;
    string academic_year = 2;
    repeated int32 teacher_ids = 3;           // optional filter
}

message AllPreferencesResponse {
    repeated TeacherPreferencesSet preferences_sets = 1;
}

message TeacherPreferencesSet {
    int32 teacher_id = 1;
    string teacher_name = 2;
    int32 teacher_priority = 3;
    repeated TeacherPreference preferences = 4;
}

// --- GROUPS ---

message CreateGroupRequest {
    string name = 1;
    string short_name = 2;
    
    int32 year = 3;
    int32 semester = 4;
    
    string program_code = 5;
    string program_name = 6;
    string specialization = 7;
    string level = 8;
    
    int32 curator_teacher_id = 9;
    string enrollment_date = 10;
}

message GetGroupRequest {
    oneof identifier {
        int32 id = 1;
        string name = 2;
    }
}

message UpdateGroupRequest {
    int32 id = 1;
    
    string name = 2;
    int32 semester = 3;
    int32 curator_teacher_id = 4;
    bool is_active = 5;
}

message DeleteGroupRequest {
    int32 id = 1;
}

message ListGroupsRequest {
    int32 page = 1;
    int32 page_size = 2;
    
    int32 year = 3;
    string level = 4;
    bool only_active = 5;
    
    string sort_by = 6;
}

message GroupResponse {
    Group group = 1;
    string message = 2;
}

message GroupsListResponse {
    repeated Group groups = 1;
    int32 total_count = 2;
}

// --- STUDENTS ---

message CreateStudentRequest {
    string full_name = 1;
    string first_name = 2;
    string last_name = 3;
    string middle_name = 4;
    
    string student_number = 5;
    int32 group_id = 6;
    
    string email = 7;
    string phone = 8;
    
    string enrollment_date = 9;
}

message GetStudentRequest {
    oneof identifier {
        int32 id = 1;
        string student_number = 2;
        int32 user_id = 3;
    }
}

message UpdateStudentRequest {
    int32 id = 1;
    
    string full_name = 2;
    int32 group_id = 3;
    string email = 4;
    string phone = 5;
    string status = 6;
}

message DeleteStudentRequest {
    int32 id = 1;
}

message ListStudentsRequest {
    int32 page = 1;
    int32 page_size = 2;
    
    int32 group_id = 3;
    string status = 4;
    
    string sort_by = 5;
}

message GroupStudentsRequest {
    int32 group_id = 1;
    string status = 2;
}

message StudentResponse {
    Student student = 1;
    string message = 2;
}

message StudentsListResponse {
    repeated Student students = 1;
    int32 total_count = 2;
}

// --- DISCIPLINES ---

message CreateDisciplineRequest {
    string name = 1;
    string short_name = 2;
    string code = 3;
    
    string department = 4;
    int32 credit_units = 5;
    string discipline_type = 6;
}

message GetDisciplineRequest {
    oneof identifier {
        int32 id = 1;
        string code = 2;
        string name = 3;
    }
}

message ListDisciplinesRequest {
    int32 page = 1;
    int32 page_size = 2;
    
    string department = 3;
    bool only_active = 4;
}

message DisciplineResponse {
    Discipline discipline = 1;
    string message = 2;
}

message DisciplinesListResponse {
    repeated Discipline disciplines = 1;
    int32 total_count = 2;
}

// --- COURSE LOADS ---

message CreateCourseLoadRequest {
    string discipline_name = 1;
    string discipline_code = 2;
    int32 discipline_id = 3;              // Опционально
    
    int32 teacher_id = 4;                 // Опционально (0 = NULL)
    int32 group_id = 5;                   // Опционально (0 = NULL)
    
    string lesson_type = 6;
    int32 hours_per_semester = 7;
    int32 weeks_count = 8;                // Опционально (default: 16)
    
    int32 semester = 9;
    string academic_year = 10;
    
    string required_classroom_type = 11;  // Опционально
    int32 min_classroom_capacity = 12;    // Опционально
}

message GetCourseLoadRequest {
    int32 id = 1;
}

message ListCourseLoadsRequest {
    int32 page = 1;
    int32 page_size = 2;
    
    int32 semester = 3;
    string academic_year = 4;
    
    repeated int32 teacher_ids = 5;
    repeated int32 group_ids = 6;
    repeated string lesson_types = 7;
    
    bool only_active = 8;
}

message DeleteCourseLoadRequest {
    int32 id = 1;
}

message CourseLoadResponse {
    CourseLoad course_load = 1;
    string message = 2;
}

message CourseLoadsListResponse {
    repeated CourseLoad course_loads = 1;
    int32 total_count = 2;
}

// --- IMPORT ---

message ImportRequest {
    bytes file_data = 1;
    string filename = 2;
    
    int32 semester = 3;
    string academic_year = 4;
    
    bool validate_only = 5;                    // Только валидация, не сохранять
    int32 imported_by = 6;
}

message ImportResponse {
    bool success = 1;
    string batch_id = 2;
    
    int32 total_rows = 3;
    int32 successful_rows = 4;
    int32 failed_rows = 5;
    repeated string errors = 6;
    
    string message = 7;
}

message ImportStatusRequest {
    string batch_id = 1;
}

message ImportStatusResponse {
    ImportBatch batch = 1;
}

message ImportBatchesRequest {
    int32 limit = 1;
    string status = 2;
}

message ImportBatchesResponse {
    repeated ImportBatch batches = 1;
}

// --- USER LINKS ---

message LinkRequest {
    int32 entity_id = 1;                       // teacher_id или student_id
    int32 user_id = 2;
}

message LinkResponse {
    bool success = 1;
    string message = 2;
}

message UserIdRequest {
    int32 user_id = 1;
}

message DeleteResponse {
    bool success = 1;
    string message = 2;
}

// --- HEALTH CHECK ---

message HealthCheckRequest {}

message HealthCheckResponse {
    string status = 1;
    string version = 2;
    string timestamp = 3;
}

